pyomo/contrib/mindtpy/feasibility_pump.py

```diff
@@ -0,0 +1,330 @@
+                solve_data.LB, solve_data.UB))
+
+    if solve_data.solution_improved:
+        solve_data.best_solution_found = solve_data.working_model.clone()
+        assert is_feasible(solve_data.best_solution_found, config), \
+            "Best found model infeasible! There might be a problem with the precisions - the feaspump seems to have converged (error**2 <= integer_tolerance). " \
+            "But the `is_feasible` check (error <= constraint_tolerance) doesn't work out"
+
+    # Always add the oa cut
+    copy_var_list_values(sub_nlp.MindtPy_utils.variable_list,
+                         solve_data.mip.MindtPy_utils.variable_list,
+                         config, ignore_integrality=True)
+    add_oa_cuts(solve_data.mip, dual_values, solve_data, config)
+
+    config.call_after_subproblem_feasible(sub_nlp, solve_data)
+
+
+def feasibility_pump_loop(solve_data, config):
+    """
+    Main loop for MindtPy Algorithms
+
+    This is the outermost function for the algorithms in this package; this function controls the progression of
+    solving the model.
+
+    Parameters
+    ----------
+    solve_data: MindtPy Data Container
+        data container that holds solve-instance data
+    config: ConfigBlock
+        contains the specific configurations for the algorithm
+    """
+    working_model = solve_data.working_model
+    main_objective = next(
+        working_model.component_data_objects(Objective, active=True))
+    while solve_data.mip_iter < config.iteration_limit:
+
+        config.logger.info(
+            '---MindtPy Master Iteration %s---'
+            % solve_data.mip_iter)
+
+        solve_data.mip_subiter = 0
+        # solve MILP master problem
+        feas_mip, feas_mip_results = solve_OA_master(
+            solve_data, config)
+        if feas_mip_results.solver.termination_condition is tc.optimal:
+            handle_master_mip_optimal(feas_mip, solve_data, config)
+        elif feas_mip_results.solver.termination_condition is tc.infeasible:
+            # This basically means the incumbent is the optimal solution
+            if solve_data.best_solution_found is not None:
+                config.logger.info(
+                    'Problem became infeasible. This means the feasibility pump has converged.')
+                solve_data.results.solver.termination_condition = tc.optimal
+            else:
+                config.logger.info('No feasible solution has been found')
+                solve_data.results.solver.termination_condition = tc.infeasible
+            break
+        elif feas_mip_results.solver.termination_condition is tc.maxIterations:
+            config.logger.error('No feasible solution has been found')
+            solve_data.results.solver.termination_condition = tc.maxIterations
+            break
+
+        # Solve NLP subproblem
+        # The constraint linearization happens in the handlers
+        fixed_nlp, fixed_nlp_result = solve_feas_pump_NLP_subproblem(
```

Comment:
inside of this function you have a way to handle the infeasible NLP subproblem with an assert, but then you call the functions below again. Is this right?